iroh/src/baomap/readonly_mem.rs

```diff
@@ -0,0 +1,293 @@
+    },
+    util::progress::{IdGenerator, ProgressSender},
+    Hash, IROH_BLOCK_SIZE,
+};
+use range_collections::RangeSet2;
+use tokio::sync::mpsc;
+
+use super::flatten_to_io;
+
+/// A readonly in memory database for iroh-bytes.
+///
+/// This is basically just a HashMap, so it does not allow for any modifications
+/// unless you have a mutable reference to it.
+///
+/// It is therefore useful mostly for testing and sharing static data.
+#[derive(Debug, Clone, Default)]
+pub struct Store(Arc<HashMap<Hash, (PreOrderMemOutboard, Bytes)>>);
+
+impl<K, V> FromIterator<(K, V)> for Store
+where
+    K: Into<String>,
+    V: AsRef<[u8]>,
+{
+    fn from_iter<T: IntoIterator<Item = (K, V)>>(iter: T) -> Self {
+        let (db, _m) = Self::new(iter);
+        db
+    }
+}
+
+impl Store {
+    /// Create a new [Store] from a sequence of entries.
+    ///
+    /// Returns the database and a map of names to computed blake3 hashes.
+    /// In case of duplicate names, the last entry is used.
+    pub fn new(
+        entries: impl IntoIterator<Item = (impl Into<String>, impl AsRef<[u8]>)>,
+    ) -> (Self, BTreeMap<String, blake3::Hash>) {
+        let mut names = BTreeMap::new();
+        let mut res = HashMap::new();
+        for (name, data) in entries.into_iter() {
+            let name = name.into();
+            let data: &[u8] = data.as_ref();
+            // compute the outboard
+            let (outboard, hash) = bao_tree::io::outboard(data, IROH_BLOCK_SIZE);
+            // add the name, this assumes that names are unique
+            names.insert(name, hash);
+            // wrap into the right types
+            let outboard =
+                PreOrderMemOutboard::new(hash, IROH_BLOCK_SIZE, outboard.into()).unwrap();
+            let data = Bytes::from(data.to_vec());
+            let hash = Hash::from(hash);
+            res.insert(hash, (outboard, data));
+        }
+        (Self(Arc::new(res)), names)
+    }
+
+    /// Insert a new entry into the database, and return the hash of the entry.
+    ///
+    /// If the database was shared before, this will make a copy.
+    pub fn insert(&mut self, data: impl AsRef<[u8]>) -> Hash {
+        let inner = Arc::make_mut(&mut self.0);
+        let data: &[u8] = data.as_ref();
+        // compute the outboard
+        let (outboard, hash) = bao_tree::io::outboard(data, IROH_BLOCK_SIZE);
+        // wrap into the right types
+        let outboard = PreOrderMemOutboard::new(hash, IROH_BLOCK_SIZE, outboard.into()).unwrap();
+        let data = Bytes::from(data.to_vec());
+        let hash = Hash::from(hash);
+        inner.insert(hash, (outboard, data));
+        hash
+    }
+
+    /// Get the bytes associated with a hash, if they exist.
+    pub fn get(&self, hash: &Hash) -> Option<Bytes> {
+        let entry = self.0.get(hash)?;
+        Some(entry.1.clone())
+    }
+
+    fn export_sync(
+        &self,
+        hash: Hash,
+        target: PathBuf,
+        _mode: ExportMode,
```

Comment:
no mode and no progress check?

Reply:
mode: see above. progress: to support progress you would no longer be able to use std::fs::write.

Reply:
Doesn't progress at least give start and end? which we could easily emit at least. Also pretty sure `fs::write` is just `File::open` + `write_all`, which should support progress

Reply:
Yeah, it was just me being lazy. Initially I did not support export at all for this store type, but then I added it.